src/map.cpp

```diff
@@ -499,89 +539,244 @@ const vehicle *map::vehproceed()
+
+        const auto &tr = ter_at( pp );
+        // Deep water and air
+        if( tr.has_flag( TFLAG_DEEP_WATER ) ) {
+            submerged_wheels++;
+            // No traction from wheel in water
+            continue;
+        } else if( tr.has_flag( TFLAG_NO_FLOOR ) ) {
+            // Ditto for air, but with no submerging
+            continue;
+        }
+
+        const int move_mod = move_cost_ter_furn( pp );
+        if( move_mod == 0 ) {
+            // Vehicle locked in wall
+            // Shouldn't happen, but does
+            return 0.0f;
+        } else {
+            traction_wheel_area += 2 * wheel_area / move_mod;
+        }
+    }
+
+    // Submerged wheels threshold is 2/3.
+    if( num_wheels > 0 && (float)submerged_wheels / num_wheels > .666) {
+        return -1;
+    }
+
+    if( total_wheel_area <= 0.01f ) {
+        debugmsg( "%s has wheel area <0.01", veh.name.c_str() );
+        return 0.0f;
+    }
+
+    return traction_wheel_area / total_wheel_area;
+}
+
+float map::vehicle_buoyancy( vehicle &veh ) const
+{
+    const tripoint pt = veh.global_pos3();
+    const auto &float_indices = veh.floating;
+    const int num = float_indices.size();
+    int moored = 0;
+    for( int w = 0; w < num; w++ ) {
+        const int p = float_indices[w];
+        const tripoint pp = pt + veh.parts[p].precalc[0];
+
+        if( !has_flag( "SWIMMABLE", pp ) ) {
+            moored++;
+        }
+    }
+
+    if( moored > num - 1 ) {
+        return 0.0f;
+    }
+
+    // TODO: Actually implement buoyancy
+    return 1.0f;
+}
+
 void map::move_vehicle( vehicle &veh, const tripoint &dp, const tileray &facing )
 {
+    const bool vertical = dp.z != 0;
+    if( ( dp.x == 0 && dp.y == 0 && dp.z == 0 ) ||
+        ( abs( dp.x ) > 1 || abs( dp.y ) > 1 || abs( dp.z ) > 1 ) ||
+        ( vertical && ( dp.x != 0 || dp.y != 0 ) ) ) {
+        debugmsg( "move_vehicle called with %d,%d,%d displacement vector", dp.x, dp.y, dp.z );
+        return;
+    }
+
+    if( dp.z + veh.smz < -OVERMAP_DEPTH || dp.z + veh.smz > OVERMAP_HEIGHT ) {
+        return;
+    }
+
     tripoint pt = veh.global_pos3();
-    bool can_move = true;
-    // Calculate parts' mount points @ next turn (put them into precalc[1])
     veh.precalc_mounts( 1, veh.skidding ? veh.turn_dir : facing.dir() );
 
     int impulse = 0;
 
-    std::vector<veh_collision> veh_veh_colls;
-    std::vector<veh_collision> veh_misc_colls;
+    std::vector<veh_collision> collisions;
 
-    if( veh.velocity == 0 ) {
-        can_move = false;
-    }
     // Find collisions
     // Velocity of car before collision
-    const int velocity_before = veh.velocity;
-    veh.collision( veh_veh_colls, veh_misc_colls, dp.x, dp.y, can_move, impulse );
+    // Split into vertical and horizontal movement
+    const int velocity_before = vertical ? veh.vertical_velocity : veh.velocity;
+    if( velocity_before == 0 ) {
+        debugmsg( "%s tried to move with no velocity", veh.name.c_str() );
```

Comment:
The message can be "easily" triggered with this: spawns a car on z-level 0. Spawn another car on top of it (or drive it there, from a debug spawned road on z-level 1). The car below counts as derivable  terrain, so it's possible to drive onto it. Make sure the above car rests only on the other car, not on any terrain below. Then go into the lower car and drive it away. The upper car will fall down (which is quite impressive), but the message appears.

Btw. Driving from on-zlevel to the below is also quite amazing.